src/storage.rs

```diff
@@ -68,11 +77,12 @@ impl<'a, S: ReadonlyStorage> ReadonlyStorage for StorageTransaction<'a, S> {
 
 impl<'a, S: ReadonlyStorage> Storage for StorageTransaction<'a, S> {
     fn set(&mut self, key: &[u8], value: &[u8]) {
-        self.local_state.set(key, value);
-        self.rep_log.push(Op::Set {
+        let op = Op::Set {
             key: key.to_vec(),
             value: value.to_vec(),
-        })
+        };
+        op.apply(&mut self.local_state);
```

Comment:
I like this, use same `apply` both for local cache and later when committing the `rep_log`